File: frontend/src/types/document.ts
/**
 * Document and Source TypeScript types.
 */

export type SourceType = 'pdf' | 'url' | 'youtube' | 'github' | 'notion' | 'text' | 'docx';

export type IngestionStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface DocumentItem {
  id: string;
  workspace_id: string;
  filename: string;
  source_type: SourceType;
  status: IngestionStatus;
  url?: string | null;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SourceItem {
  id: string;
  workspace_id: string;
  type: SourceType;
  title: string;
  url?: string | null;
  content_preview?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
}

export interface DocumentCreatePayload {
  source_type: SourceType;
  filename: string;
  url?: string;
  content?: string;
  metadata?: Record<string, unknown>;
}

File: backend/app/db/models/document.py
"""Document model — an uploaded or linked knowledge source file."""

import uuid

from sqlalchemy import ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, UUIDMixin


class Document(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "documents"

    workspace_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("workspaces.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    filename: Mapped[str] = mapped_column(String(512), nullable=False)
    mime_type: Mapped[str] = mapped_column(String(128), nullable=False)
    storage_key: Mapped[str | None] = mapped_column(String(1024), nullable=True)
    source_type: Mapped[str] = mapped_column(
        String(50), nullable=False, default="file"
    )  # file, url, youtube, github
    status: Mapped[str] = mapped_column(
        String(50), nullable=False, default="pending"
    )  # pending, processing, completed, failed
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB, nullable=True)
    content: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relationships
    workspace = relationship("Workspace", back_populates="documents")
    chunks = relationship("DocumentChunk", back_populates="document", cascade="all, delete-orphan")

    def __repr__(self) -> str:
        return f"<Document {self.filename} ({self.status})>"
